internal/training/grpo_engine/policy.c

```
@@ -0,0 +1,874 @@
+    if (!t_embed || !t_norm || !t_output) {
+        fprintf(stderr, "policy: missing global tensors (embed=%p, norm=%p, output=%p)\n",
+                (void*)t_embed, (void*)t_norm, (void*)t_output);
+        munmap(pe->mmap_base_actual, pe->mmap_size);
+        gguf_close(&pe->gf);
+        free(pe);
+        return NULL;
+    }
+
+    pe->embed = (const void *)((const uint8_t *)pe->mmap_base + (t_embed->offset - pe->gf.data_offset));
+    pe->embed_dtype = (int)t_embed->dtype;
+    pe->output_norm = (const float *)((const uint8_t *)pe->mmap_base + (t_norm->offset - pe->gf.data_offset));
```

Cppcheck (2.21.0)

[portability] Casting between const unsigned char * and const float * which have an incompatible binary data representation.

(invalidPointerCast)